src/NuGet.Core/NuGet.Packaging/Signing/Verification/SignatureTrustAndValidityVerificationProvider.cs

```diff
@@ -26,45 +25,109 @@ public SignatureTrustAndValidityVerificationProvider()
+            var status = primarySignatureVerificationSummary.Status;
+
+            if (primarySignatureHasCountersignature)
+            {
+                if (settings.AlwaysVerifyCountersignature || ShouldFallbackToRepositoryCountersignature(primarySignatureVerificationSummary))
+                {
+                    var counterSignature = RepositoryCountersignature.GetRepositoryCountersignature(signature);
+                    verifySettings = new SignatureVerifySettings(
+                        treatIssuesAsErrors: !settings.AllowIllegal,
+                        allowUntrustedRoot: false,
+                        allowUnknownRevocation: settings.AllowUnknownRevocation,
+                        logOnSignatureExpired: true);
+
+                    var countersignatureVerificationSummary = VerifyValidityAndTrust(counterSignature, settings, verifySettings, certificateExtraStore, issues);
+                    status = countersignatureVerificationSummary.Status;
+
+                    if (!Rfc3161TimestampVerificationUtility.ValidateSignerCertificateAgainstTimestamp(signature.SignerInfo.Certificate, countersignatureVerificationSummary.Timestamp))
+                    {
+                        issues.Add(SignatureLog.Issue(!settings.AllowIllegal, NuGetLogCode.NU3011, string.Format(CultureInfo.CurrentCulture, Strings.SignatureNotTimeValid, nameof(AuthorPrimarySignature))));
+                        status = SignatureVerificationStatus.Illegal;
+                    }
+                }
+                else if (primarySignatureVerificationSummary.Flags.HasFlag(SignatureVerificationStatusFlags.CertificateExpired))
+                {
+                    // We are not adding this log if the primary signature has a countersignature to check the expiration against the countersignature's timestamp.
+                    // If the countersignature shouldn't be check and the primary signature was expired, add this log.
+                    issues.Add(SignatureLog.Issue(!settings.AllowIllegal, NuGetLogCode.NU3011, string.Format(CultureInfo.CurrentCulture, Strings.SignatureNotTimeValid, nameof(AuthorPrimarySignature))));
+                }
+            }
+
+            return new SignedPackageVerificationResult(status, signature, issues);
+        }
+
+        private SignatureVerificationSummary VerifyValidityAndTrust(
+            Signature signature,
+            SignedPackageVerifierSettings timestampSettings,
+            SignatureVerifySettings settings,
+            X509Certificate2Collection certificateExtraStore,
+            List<SignatureLog> issues)
         {
             var timestampIssues = new List<SignatureLog>();
 
             Timestamp validTimestamp;
             try
             {
                 validTimestamp = signature.GetValidTimestamp(
-                    settings,
+                    timestampSettings,
```

Comment:
I understand that it's being passed to a method that validates timestamps, but calling this `timestampSettings` seems odd, especially since it's just general signed package verifier settings in the caller.